src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs

```diff
@@ -171,7 +176,12 @@ void ProcessObject(JsValue valueToInsert, in string propertyAsString, IndexField
                             {
                                 value = TypeConverter.ToBlittableSupportedType(val, flattenArrays: false, forIndexing: true, canTryJsStringToDateConversion: _ticksSupport, engine: documentToProcess.Engine,
                                     context: indexContext);
-                                numberOfCreatedFields = GetRegularFields(instance, field, CreateValueForIndexing(value, propertyBoost), indexContext, sourceDocument, out _);
+
+                                //We've to wrap dynamic fields to avoid fields cache (and value override), solves: RavenDB_15983
+                                //Also backward compatibility for older indexes.
+                                numberOfCreatedFields = isDynamicFieldsEnumeratorScope && _dynamicFieldsDynamicAnalyzer
+                                    ? GetRegularFields(instance, field, LuceneCreateField(currentIndexingScope, field, CreateValueForIndexing(value, propertyBoost)), indexContext, sourceDocument, out _) 
+                                    : GetRegularFields(instance, field, CreateValueForIndexing(value, propertyBoost), indexContext, sourceDocument, out _);
```

Comment:
Can we rewrite it to:

```suggestion
object valueForIndexing;

if (isDynamicFieldsEnumeratorScope && _dynamicFieldsDynamicAnalyzer)
    valueForIndexing = LuceneCreateField(currentIndexingScope, field, CreateValueForIndexing(value, propertyBoost));
else
    valueForIndexing = CreateValueForIndexing(value, propertyBoost);
```
?